### ISSUE TITLE
Question: is it necessary to use `model.eval()`  before directly using the model?

### ISSUE THREAD

**Issue by zhong-yy** [seisbench/seisbench#186]
Hi, maybe this question sounds too simple, but I am still confused after searching some online resourses.

It seems that `model.eval()` followed by `with torch.no_grad():` is a routine step for evaluation. For example see https://pytorch-lightning.readthedocs.io/en/1.8.0/deploy/production_intermediate.html

But turning on `model.eval()` even makes the performance worse, why?

Here is an example. 

``` python
# packages
import seisbench
import seisbench.models as sbm
import seisbench.data as sbd
import seisbench.generate as sbg
import torch
import matplotlib.pyplot as plt
import numpy as np

# Data
phase_dict = {
    "trace_p_arrival_sample": "P",
    "trace_pP_arrival_sample": "P",
    "trace_P_arrival_sample": "P",
    "trace_P1_arrival_sample": "P",
    "trace_Pg_arrival_sample": "P",
    "trace_Pn_arrival_sample": "P",
    "trace_PmP_arrival_sample": "P",
    "trace_pwP_arrival_sample": "P",
    "trace_pwPm_arrival_sample": "P",
    "trace_s_arrival_sample": "S",
    "trace_S_arrival_sample": "S",
    "trace_S1_arrival_sample": "S",
    "trace_Sg_arrival_sample": "S",
    "trace_SmS_arrival_sample": "S",
    "trace_Sn_arrival_sample": "S",
}
data = sbd.ETHZ(sampling_rate=100)
data.filter(~np.isnan(data.metadata["trace_P1_arrival_sample"]),inplace=True)
generator = sbg.GenericGenerator(data)
generator.augmentation(sbg.ChangeDtype(np.float32))
generator.augmentation(sbg.WindowAroundSample(
                        list(phase_dict.keys()),
                        samples_before=3000,
                        windowlen=6000,
                        selection="random",
                        strategy="variable",))
# default amp_norm_type is "peak"
generator.augmentation(sbg.Normalize(detrend_axis=-1, amp_norm_axis=-1, amp_norm_type="peak"))
generator.augmentation(sbg.ProbabilisticLabeller(label_columns=phase_dict, sigma=5, dim=-2))

# model predictiton
idx=100
sample = generator[idx]

model = sbm.PhaseNet.from_pretrained("original")
model.cuda();
model.eval()
# model.train()
with torch.no_grad():
    pred=model(torch.tensor(sample["X"], device=model.device).unsqueeze(0))
    pred=pred[0].cpu().numpy() 
    
# plot
def plot_sample(pred,sample,fname,model_labels):
    fig,axs=plt.subplots(4,1,figsize=(7,10))
    clrs=["blue","red","green"]
    labels=["P","S","N"]
    wave_labs=["Z","N","E"]
    for j in range(3):
        axs[0].plot(sample["X"][j],label=wave_labs[j])
        axs[1].plot(sample["y"][j],label=labels[j])
        axs[2].plot(pred[model_labels.index(labels[j])],label=labels[j])
    axs[3].plot(1-pred[model_labels.index("N")],color="black",label=r"$1-noise$")
    for ax in axs: ax.legend()
    axs[1].set_ylabel("Ground True")
    axs[2].set_ylabel("Prediction")
    plt.savefig(fname=fname,bbox_inches="tight")

plot_sample(pred,sample,f"eq_x",model.labels)
```
For `model=sbm.PhaseNet.from_pretrained("original")`, turning on `model.eval()` makes the performance very bad. 
However, for `model=sbm.PhaseNet.from_pretrained("stead")`, using `model.eval()` is better. But if the data is not normalized, predicting without `model.eval()` is better.

I notice ‘model.eval()‘ is also used in [model/base.py](https://github.com/seisbench/seisbench/blob/beddc57fcec8b3602f34a57f0610fbf0ad418481/seisbench/models/base.py#L1801). Why does the above example not work in “eval” mode?

When should I use `model.eval()` before using the model?  Thank you.

**Comment by zhong-yy** [comment]
It is probably related to the normalization. If I change normalization to `generator.augmentation(sbg.Normalize(detrend_axis=-1, amp_norm_axis=-1, amp_norm_type='std'))`, the original phasenet model can work with `model.eval()`. However, in this case `sbm.PhaseNet.from_pretrained("stead")` will give weird results.

[Here](https://github.com/seisbench/seisbench/blob/beddc57fcec8b3602f34a57f0610fbf0ad418481/seisbench/models/phasenet.py#L143-L149), the PhaseNet in Seisbench indeed uses "std" as normalization type.

How come `sbm.PhaseNet.from_pretrained("original")` and `sbm.PhaseNet.from_pretrained("stead")` behave differently?

**Comment by yetinam** [comment]
Hi @zhong-yy ,

to answer your original question, you should always use `model.eval()` when using a model to predict. This puts the model into evaluation mode meaning, for example, disabling dropout or changing the behaviour of batch norm.

So why do you see different behavior for different normalization? There are two (common) ways to normalize, peak normalization and std normalization. The model needs the same normalization strategy in training and in application, otherwise performance will be degraded. The "original" model was trained with std, the other models with peak normalization. That's why you need to change this in your code.

With your analysis, you figured out an issue in the annotate step. I will open a separate bug for tracking this and fix the issue soon.

**Comment by zhong-yy** [comment]
Thank you for answering my question. That makes sense to me. @yetinam,

**Comment by yetinam** [comment]
Great! Closing this now but feel free to reopen if necessary.

And thanks again for helping us figure out the normalization issue. That really makes a difference!
